src/codewhisperer/service/keyStrokeHandler.ts

```diff
@@ -42,6 +42,49 @@ export class KeyStrokeHandler {
         return (this.#instance ??= new this())
     }
 
+    public startIdleTimeTriggerTimer(
+        event: vscode.TextDocumentChangeEvent,
+        editor: vscode.TextEditor,
+        client: DefaultCodeWhispererClient,
+        config: ConfigurationEntry
+    ) {
+        if (this.idleTriggerTimer) {
+            clearInterval(this.idleTriggerTimer)
+            this.idleTriggerTimer = undefined
+        }
+        if (!this.shouldTriggerIdleTime()) {
+            return
+        }
+        this.idleTriggerTimer = setInterval(() => {
+            const duration = (performance.now() - RecommendationHandler.instance.lastInvocationTime) / 1000
+            if (duration < CodeWhispererConstants.invocationTimeIntervalThreshold) {
+                return
+            }
+
+            try {
+                this.invokeAutomatedTrigger('IdleTime', editor, client, config)
+            } finally {
+                if (this.idleTriggerTimer) {
+                    clearInterval(this.idleTriggerTimer)
+                    this.idleTriggerTimer = undefined
+                }
+            }
+        }, CodeWhispererConstants.idleTimerPollPeriod)
+    }
+
+    public shouldTriggerIdleTime(): boolean {
+        if (isCloud9() && RecommendationHandler.instance.isGenerateRecommendationInProgress) {
```

Comment:
Does this mean in C9 we are not using idle time trigger but we removed keystroke trigger?

Reply:
in C9 we are also using idle time trigger, this condition is just to prevent concurrent triggering

Reply:
we set this flag here: https://github.com/aws/aws-toolkit-vscode/blob/e6eb99db8d7e4e8b2f825f7c89660d633053d306/src/codewhisperer/service/keyStrokeHandler.ts#L167